Fix bug lp:1521905 Avoid deadlock with backoff empty free list algorithm

ihanick opened the pull request:
by disabling it on small buffer pools

	innodb empty free list algorithm backoff requires free pages from LRU for
	best performance. buf_LRU_buf_pool_running_out cancels query if 1/4 of
       	buffer pool belongs to LRU or freelist.
	At the same time buf_flush_LRU_list_batch keep up to BUF_LRU_MIN_LEN in LRU.
       	In order to avoid deadlock baclkoff requires buffer pool to be at least
	4*BUF_LRU_MIN_LEN, but flush peformance is bad because of trashing
	and additional BUF_LRU_MIN_LEN pages are requested.

ihanick commented:
bug report: https://bugs.launchpad.net/percona-server/+bug/1521905

http://jenkins.percona.com/job/percona-server-5.6-param/1063/

laurynas-biveinis commented on plugin/tokudb-backup-plugin/Percona-TokuBackup:
```diff
@@ -1 +1 @@
-Subproject commit d90ccc177273725e9f87436bc9c3de2d05a8e1fb
+Subproject commit 1b9bb16ad74588601d8fefe46c74cc1dac1dd1d5
```
Why?

laurynas-biveinis commented on storage/tokudb/PerconaFT:
```diff
@@ -1 +1 @@
-Subproject commit 8f45478750ecb55e580dcee60f5f543848719e47
+Subproject commit cc7b1780eefa79936fd8a1d459626dddec457545
```
Why?

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -3691,6 +3691,23 @@ innobase_init(
+	/* Do not enable backoff algorithm for small buffer pool.
+	Check innodb_srv_empty_free_list_algorithm_validate
+	for additional information */
+	if ( innobase_buffer_pool_size / srv_page_size
```
This condition should be extracted to its own function, and used from both here and further in ..._validate, and the big explaining comment moved to it.

laurynas-biveinis replied:
No space after opening brace

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -3691,6 +3691,23 @@ innobase_init(
+			== SRV_EMPTY_FREE_LIST_BACKOFF) {
+		sql_print_information(
+				"InnoDB: innodb_empty_free_list_algorithm "
+				"has been changed to legacy. "
```
"... to legacy because of small buffer pool size."

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -3691,6 +3691,23 @@ innobase_init(
+				"InnoDB: innodb_empty_free_list_algorithm "
+				"has been changed to legacy. "
+				"In order to use backoff, "
+				"increase buffer pool up to 20MB.\n");
```
"... at least up to 20MB.\n"

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -3691,6 +3691,23 @@ innobase_init(
+	Check innodb_srv_empty_free_list_algorithm_validate
+	for additional information */
+	if ( innobase_buffer_pool_size / srv_page_size
+			< BUF_LRU_MIN_LEN*(4 + 1)
```
Spaces around multiplication

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -16887,6 +16904,71 @@ innodb_status_output_update(
+	ulint	algo;
+	srv_empty_free_list_t 		algorithm;
+
+	ut_a(save != NULL);
```
IMHO no need (a NULL dereference later is roughly identical to such release build assert)

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -16887,6 +16904,71 @@ innodb_status_output_update(
+	if (algo == array_elements( innodb_empty_free_list_algorithm_names) - 1)
+		return(1);
+
+	algorithm = (srv_empty_free_list_t)algo;
```
C++ style static_cast

laurynas-biveinis commented:
Please add testcases: 1) that with small bp size the startup option indeed becomes legacy; 2) likewise for dynamic setting (server started with legacy and attempt to set to backoff).

laurynas-biveinis commented:
This also needs to be done on a GCA branch for 5.6 with a null upmerge to 5.7

ihanick commented:
jenkins build after fixes:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1067/

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -1243,6 +1243,16 @@ normalize_table_name_low(
+static
+bool
+innodb_empty_free_list_algorithm_check(
+	ulong 				algorithm,
```
srv_empty_free_list_t arg type

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -1243,6 +1243,16 @@ normalize_table_name_low(
+@return	true if it's not possible to enable backoff. */
+static
+bool
+innodb_empty_free_list_algorithm_check(
```
The name should be more descriptive (what is being "check"-ed?), for example, innodb_empty_free_list_algorithm_backoff_allowed, with return condition inverted

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -1243,6 +1243,16 @@ normalize_table_name_low(
 
 /*************************************************************//**
+Empty free list algorithm.
+Checks if buffer pool is enough to enable backoff algorithm.
```
The previous descriptive comment is gone?

laurynas-biveinis commented on storage/innobase/handler/ha_innodb.cc:
```diff
@@ -16887,6 +16912,71 @@ innodb_status_output_update(
+	algorithm = static_cast<srv_empty_free_list_t>(algo);
+	if (innodb_empty_free_list_algorithm_check(algorithm,
+				innobase_buffer_pool_size / srv_page_size))
+		return(1);
```
I am starting to think that maybe sql_print_warning is required here why the variable change was forbidden?

ihanick commented:
http://jenkins.percona.com/view/PS%205.6/job/percona-server-5.6-param/1071/
- added sql_print_warning
- renamed check function, added comment
- srv_empty_free_list_t

laurynas-biveinis commented:
LGTM, please prepare a 5.7 null merge branch in order to push this.

ihanick commented:
5.7 PR: https://github.com/percona/percona-server/pull/306
